### ISSUE TITLE
kvserver: remove ability to send log entries in raft snaps

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#66834]
This is mostly dead code and has been for multiple releases - we don't send any log entries with the snapshots any more:

https://github.com/cockroachdb/cockroach/blob/1ac67a16817238f601f0b49bd7a67ad3c1531505/pkg/kv/kvserver/replica_command.go#L2438-L2446

Also, as of 21.2 (i.e. master now), we actually know for sure it is dead code:

https://github.com/cockroachdb/cockroach/blob/d546be9915c304dff495dc32b15058169b3d75e6/pkg/clusterversion/cockroach_versions.go#L233-L242

This will let us significantly simplify `kvBatchSnapshotStrategy.Send`, which can always assume that the truncated index for the snapshot matches the applied index because this branch becomes unconditional:

https://github.com/cockroachdb/cockroach/blob/1ac67a16817238f601f0b49bd7a67ad3c1531505/pkg/kv/kvserver/replica_command.go#L2448-L2462

The one weird little thing is this line,

https://github.com/cockroachdb/cockroach/blob/1ac67a16817238f601f0b49bd7a67ad3c1531505/pkg/kv/kvserver/replica_command.go#L2446

and whether it can ever be false. I don't think it can, as truncations only become visible when they are applied (at least that's how we're doing it today). So this can be turned into an `errors.AssertionFailedf`.

**Comment by erikgrinaker** [comment]
@irfansharif is doing some of this cleanup in #69887 (thanks!), but looks like we can still do some further simplifications as outlined above in a follow-up PR -- specifically, the `if canAvoidSendingLog` bit.

**Comment by tbg** [comment]
Basically the condition becomes

```go
canAvoidSendingLog := snap.State.TruncatedState.Index < snap.State.RaftAppliedIndex
```

I think this was "just" FUD when I wrote that code, but there's a potential for a real problem there. However, the problem wouldn't be around sending snapshots, Raft would just explode out from under us.

Take a Raft instance with log `[10, 1000]` (all committed), but it is currently catching up on entry application, say the state machine is at entry 20 and will consume 10 more entries per raft cycle (we do paginate the entry commit to avoid pulling too much into memory at once). A log truncation could arrive suggesting to cut off the log at index, say, 500. So now we've made it impossible to move forward the state machine using the log, oops. I filed https://github.com/cockroachdb/cockroach/issues/70109 for this. But we can hard-code `canAvoidSendingLog` to true and make the resulting simplifications.
